client/wsclient.go

```diff
@@ -245,19 +250,33 @@ func (c *wsClient) runOneCycle(ctx context.Context) {
 		c.common.PackagesStateProvider,
 		c.common.Capabilities,
 	)
-	r.ReceiverLoop(ctx)
-
-	// Stop the background processors.
-	procCancel()
-
-	// If we exited receiverLoop it means there is a connection error, we cannot
-	// read messages anymore. We need to start over.
+	r.Start(ctx)
+
+	select {
+	case <-c.sender.IsStopped():
+		// sender will send close message to initiate the close handshake
+		if err := c.sender.Err(); err != nil && err != websocket.ErrCloseSent {
+			c.common.Logger.Debugf("failed to send close message, close without the handshake.")
+			break
+		}
 
-	// Close the connection to unblock the WSSender as well.
-	_ = c.conn.Close()
+		c.common.Logger.Debugf("waiting for close message from server.")
+		select {
+		case <-r.IsStopped():
+			c.common.Logger.Debugf("shutdown handshake complete.")
+		case <-time.After(c.connShutdownTimeout):
+			c.common.Logger.Debugf("timeout waiting for close message.")
+			// not receive close message from the server, close the connection to force the receive loop to stop
+			_ = c.conn.Close()
+			<-r.IsStopped()
+		}
+	case <-r.IsStopped():
+		// If we exited receiverLoop it means there is a connection error, we cannot
+		// read messages anymore. We need to start over.
 
-	// Wait for WSSender to stop.
-	c.sender.WaitToStop()
+		stopSender()
+		<-c.sender.IsStopped()
```

Comment:
Does this waiting require a timeout?

Reply:
I don't think it's necessary, as mentioned above(https://github.com/open-telemetry/opamp-go/pull/213/files#r1446914020)

Reply:
I am not sure about it. 

Calling `stopSender()` here triggers context cancellation. That in turn causes the `sender` to call `sendCloseMessage()`. `sendCloseMessage()` can be blocked indefinitely if the peer server is not reading from the socket and there is no room in the local socket buffer. This means the `sender` will never be stopped and `IsStopped()` channel will never be closed. We can block indefinitely long here.

I think we need a timeout here and need to close the connection forcedly like we do it to force the receiver to stop above in line 270.

I may be missing something here. Please help me with the analysis.

Reply:
OK, from the comment thread `sendCloseMessage()` will not blocked indefinitely, it will return at most after `defaultSendCloseMessageTimeout` so this does not need a timeout.